### PR TITLE
change dir_is_tagged to use os.path.exists()

### PR DESCRIPTION
**Pull request by costela**
Very small commit to make tagging accept folders as tags and allow things like
`--exclude-if-present .git`.
This is being discussed in #1999

### CODE REVIEW

**Comment by ThomasWaldmann** [comment]
Hmm, 2 tests broken currently.

**Comment by costela** [comment]
Oops, sorry about that. Should be passing now (the Xcode tests are still in the queue).

File: src/borg/testsuite/archiver.py [hunk 1]
```diff
@@ -932,16 +932,20 @@ def _create_test_keep_tagged(self):
         self.create_regular_file('taggedall/%s' % CACHE_TAG_NAME,
                                  contents=CACHE_TAG_CONTENTS + b' extra stuff')
         self.create_regular_file('taggedall/file4', size=1024)
+        self.create_regular_file('taggeddir/.NOBACKUP2/file5', size=1024)
+        self.create_regular_file('taggeddir/file6', size=1024)
```
Comment (ThomasWaldmann) [inline src/borg/testsuite/archiver.py]: as we now consider tags no matter whether they are files or dirs, maybe just modify line925 to cover the dirs case (and drop "taggeddir")?

tagged1..n dirs were meant to cover the n different ways to tag.

taggedall was meant to have all tagging methods present at once, maybe also modify that accordingly.

**Comment by ThomasWaldmann** [comment]
Hmm, 1.0-maint or master?

**Comment by enkore** [comment]
master due to changed behaviour?

**Comment by ThomasWaldmann** [comment]
Guess we must use master (and introduce it in 1.1), because in 1.0 it is documented specifically as acting on **files**:
```
  --exclude-if-present FILENAME
                        exclude directories that contain the specified file
  --keep-tag-files      keep tag files of excluded caches/directories
 ```
Also, these docs need to be updated so they do not just refer to "tag files", "FILENAME", "specified file", ...

**Comment by costela** [comment]
@ThomasWaldmann, sorry for the delayed reaction. Just pushed a new commit with the suggested changes to testing + small corrections to the help texts. Probably not optimal, but still better than nothing! ;)

File: src/borg/archiver.py [hunk 2]
```diff
@@ -1962,7 +1962,7 @@ def build_parser(self, prog=None):
                                    help='exclude directories that contain the specified file')
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: also change above?

File: src/borg/archiver.py [hunk 3]
```diff
@@ -2550,7 +2550,7 @@ def build_parser(self, prog=None):
...
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: also change above?

**Comment by costela** [comment]
Last two suggestions incorporated (plus a minor clarification for the `--keep-tag-files` option).
Still a bit unfortunate that the 'files' in the option's name doesn't quite reflect the option's semantics, but I guess it's not worth some API-braking rename or the introduction of a new option.

**Comment by ThomasWaldmann** [comment]
@costela yes, i noticed the option name issue also.

We could improve that like this:
- invent a good, additional option name that triggers the new behaviour
- keep the old option and keep its old behaviour, but deprecate it and issue a deprecation warning, so we can remove it (in 1.1, in 1.2 or whatever future version)

**Comment by enkore** [comment]
Seems like an awful lot of hassle for such a small change.

**Comment by costela** [comment]
@enkore how about a compromise?
Keep the new behavior, but add a new `--keep-tag` option that acts as an alias to `--keep-tag-file`. Deprecate `--keep-tag-file` and remove it some time in the future. This way the slight semantic problem gets fixed eventually, without having to add too much code to this minor change.

**Comment by enkore** [comment]
ok

I mainly don't want to have all the code duplicated.

**Comment by ThomasWaldmann** [comment]
@costela sounds good. there is already some code for dealing with option deprecations, use it.

**Comment by costela** [comment]
Another attempt: I took the liberty of bikeshedding a bit further and using the name `--keep-exclude-tags`, also making the help text a bit more explicit. I also added some testing for the deprecation.

File: src/borg/archiver.py [hunk 4]
```diff
@@ -1959,10 +1960,11 @@ def build_parser(self, prog=None):
                                         'http://www.brynosaurus.com/cachedir/spec.html)')
         exclude_group.add_argument('--exclude-if-present', dest='exclude_if_present',
                                    metavar='FILENAME', action='append', type=str,
```
Comment (ThomasWaldmann) [inline src/borg/archiver.py]: this still says FILENAME. maybe rather PATH?

Comment (costela) [inline src/borg/archiver.py]: ah, good catch!

File: docs/changes.rst [hunk 5]
```diff
@@ -175,6 +175,9 @@ Other changes:
 - make RPC protocol more extensible: use named parameters.
 - RemoteRepository: misc. code cleanups / refactors
 - clarify cache/repository README file
+- the --keep-tag-files option has been deprecated in favor of --keep-exclude-tags.
+  additionally, both options now also accept directories as tags for backup
+  exclusion, #1999
```
Comment (ThomasWaldmann) [inline docs/changes.rst]: this should first talk about --exclude-if-present now just checking for the existance of a filesystem object of that name (not requiring the tag to be a file, but also accepting e.g. a directory).
the keeping behaviour should be described afterwards.

File: docs/usage/create.rst.inc [hunk 6]
```diff
@@ -36,10 +36,10 @@ Exclusion options
         | read exclude patterns from EXCLUDEFILE, one per line
     ``--exclude-caches``
         | exclude directories that contain a CACHEDIR.TAG file (http://www.brynosaurus.com/cachedir/spec.html)
-    ``--exclude-if-present FILENAME``
-        | exclude directories that contain the specified file
-    ``--keep-tag-files``
-        | keep tag files of excluded caches/directories
+    ``--exclude-if-present PATH``
```
Comment (ThomasWaldmann) [inline docs/usage/create.rst.inc]: while writing above comment, i noticed i was wrong suggesting the usage of "PATH" here, sorry.
it must not be a path (like /a/b/c), just a NAME.

File: docs/usage/create.rst.inc [hunk 7]
```diff
@@ -36,10 +36,10 @@ Exclusion options
...
+        | exclude directories that contain the specified file/directory
```
Comment (ThomasWaldmann) [inline docs/usage/create.rst.inc]: ... that are tagged by containing a filesystem object of the given NAME

File: docs/usage/create.rst.inc [hunk 8]
```diff
@@ -36,10 +36,10 @@ Exclusion options
...
+    ``--keep-exclude-tags``
+        | keep tag files/directories (i.e.: arguments to --exclude-if-present) in otherwise excluded caches/directories
```
Comment (ThomasWaldmann) [inline docs/usage/create.rst.inc]: keep tag objects...

File: docs/usage/recreate.rst.inc [hunk 9]
```diff
@@ -36,10 +36,10 @@ Exclusion options
...
-        | exclude directories that contain the specified file
-    ``--keep-tag-files``
-        | keep tag files of excluded caches/directories
+    ``--exclude-if-present PATH``
+        | exclude directories that contain the specified file/directory
+    ``--keep-exclude-tags``
+        | keep tag files/directories (i.e.: arguments to --exclude-if-present) in otherwise excluded caches/directories
 
 Archive options
     ``--target TARGET``
```
Comment (ThomasWaldmann) [inline docs/usage/recreate.rst.inc]: maybe do not update the .rst.inc file at all, these files are usually created / updated at release time.

changes in here also often cause merge conflicts e.g. when merging one branch into another.

File: docs/changes.rst (hunk 5 above)
Comment (ThomasWaldmann) [inline docs/changes.rst]: also, "both options now also accept directories as tags" is a bit misleading in the context of "--keep-tag-files option" and "--keep-exclude-tags" - these options do not get any NAME as value, but are just booleans.

so, this refers to the --exclude-if-present NAME options rather.

**Comment by ThomasWaldmann** [comment]
oh, the changelog entry should be replaced into b4 section, b3 is already released.

**Comment by costela** [comment]
Incorporated latest suggestions and rebased on latest master. Now let's hope nothing breaks ;)

**Pull request merged**